auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java

```diff
@@ -244,6 +253,114 @@ public void onClick(View view) {
         }
     }
 
+    /**
+     * This function is very similar with poplulateIdpList, but specifically made for custom layout
+     */
+    private void populateIdpListCustomLayout(List<IdpConfig> providerConfigs,
+                                             final SocialProviderResponseHandler handler) {
+        ViewModelProvider supplier = ViewModelProviders.of(this);
+
+        Map<String, Integer> providerButtonIds = customLayout.getProvidersButton();
```

Comment:
This is a lot of duplicated code with the non-custom-layout version.  It seems the only difference is how to get the button ID for a provider?
 
Also we should still be using FlowParameters as the official list of which providers were configured, not the list of buttons provided in the layout object.  If there's a provider configured in `FlowParameters` that does  not have a custom button, we should throw an exception.

Reply:
Yes. I notice that as well. 

I'm still thinking whether to join them as one function, or just make them separately. Making it separated would be better idea (I think) as it would be easier for future improvements.

Reply:
> Also we should still be using FlowParameters as the official list of which providers were configured, not the list of buttons provided in the layout object. If there's a provider configured in FlowParameters that does not have a custom button, we should throw an exception.

And yes. I'll make sure of this as well.